forcetree.c

```diff
@@ -221,152 +313,143 @@ int force_tree_build_single(int npart)
      * helps cache locality. In my tests without this list the
      * reduction in cache performance destroyed the benefit of
      * parallelizing this loop!*/
-    int nfree_thread=nfree;
-    int numfree_thread=0;
+    int nnext_thread=nnext;
+    int nrem_thread=0;
+    /* Stores the last-seen node on this thread.
+     * Since most particles are close to each other, this should save a number of tree walks.*/
+    int this_acc = firstnode;
     /* now we insert all particles */
-    /*Initialise the Morton Keys to save time later.*/
-    morton_t * MortonKey = mymalloc("MortonKeys",npart*sizeof(morton_t));
-    #pragma omp parallel for
-    for(i=0; i < npart; i++) {
-           MortonKey[i] = MORTON(P[i].Pos);
-    }
 #ifdef OPENMP_USE_SPINLOCK
-    /*Initialise some spinlocks*/
-    pthread_spinlock_t * SpinLocks = mymalloc("NodeSpinlocks",MaxNodes*sizeof(pthread_spinlock_t));
-    for(i=0; i<nfree - All.MaxPart; i++) {
-            pthread_spin_init(&SpinLocks[i], 0);
+    /*Initialise some spinlocks off*/
+    pthread_spinlock_t * SpinLocks = mymalloc("NodeSpinlocks", (lastnode - firstnode)*sizeof(pthread_spinlock_t));
+    for(i=0; i < lastnode - firstnode; i++) {
+        pthread_spin_init(&SpinLocks[i],PTHREAD_PROCESS_PRIVATE);
     }
-
-    #pragma omp parallel for firstprivate(nfree_thread, numfree_thread)
+    #pragma omp parallel for firstprivate(nnext_thread, nrem_thread, this_acc)
 #endif
     for(i = 0; i < npart; i++)
     {
         /*Can't break from openmp for*/
-        if(nfree_thread >= All.MaxPart + MaxNodes-1)
+        if(nnext_thread >= lastnode-1)
             continue;
 
         /*First find the Node for the TopLeaf */
 
-        int shift;
-        int topleaf = domain_get_topleaf_with_shift(P[i].Key, &shift);
-        int this = TopLeaves[topleaf].treenode;
+        int this;
+        if(inside_node(&Nodes[this_acc], i)) {
```

Comment:
Did you finally get rid of 'shift'. I know it is implied during the tree descent but dared not touch it.. How did it work out for you?

Reply:
I got rid of it - I got rid of the Morton key stuff entirely and just compared the particle position to the node center directly. It turned out that computing the Morton keys was much slower than the speedup from using them. So the tree build got faster by a factor two without them.

Reply:
This is very surprising. Were you recomputing the morton key before the comparison? the morton key based approach is usually believed to be much faster, since it is only a few bit operations; the position is a bunch of floats and touches more memory bytes.

Reply:
The Morton Key was computed in a separate loop over each particle, once per tree walk, and stored in temporary memory. But I found that this was fairly slow, and computation of the Morton Key was fairly slow, and the overhead from it dominated the comparison time. 

I suspect this is a consequence of processor prefetching - at the time when the subnode is being computed the particle structure is already in cache because we read P[i].Key and this brings in P[i].Pos as well. But Morton[i] is stored elsewhere and was not prefetched, so the speedup from it being fewer bits was small.

I also thought the absolute fastest way to do this would be to use the peano-hilbert-key, which is pre-computed already and already in cache. But I could not figure out how to do this - it requires knowing the rotation at all levels.